### PR TITLE
VIM-7331: Move Google SignIn Helpers into Shared Library

### PR DESCRIPTION
**Pull request by nicolelehrer**
#### Ticket

[VIM-7331](https://vimean.atlassian.net/browse/VIM-7331)

#### Pull Request Checklist

- [X] Resolved any merge conflicts
- [X] No build errors or warnings are introduced
- [X] New files are written in Swift
- [X] New classes contain license headers
- [X] New classes have Documentation
- [X] New public methods have Documentation

#### Issue Summary

- Adds Google Sign in API (moved from iOS client)
- Breaking changes for client on related public interfaces

#### Implementation Summary

- `AuthenticationController`: Adds authentication methods for Google Login and Join. Two changes from previous implementation 
   - we now inject the scopes 
   - the `marketingOptIn` parameter across all authentication methods is always passed in as a Bool, and then converted to a String (the expected format by the Vimeo API) inside of the request builder 
- `Constants`: Added an enum `RawTypeConverter` which intentionally wraps the conversion of any raw types to those expected by the API, such as Bool to String "false" or "true". While we could easily do this in line using `String(aBool)` or `aBool.description`, because bad request parameter serialization can be hard to diagnose, I preferred that there would be a single point of conversion that 
    - (1) does not rely on the fact that the expected type just so happens to match how Swift converts `Bool` to `String` and 
    - (2) can easily be updated if requirements change
- `RawTypeConverterTests`: tests the above
- `Request+Authentication`: 
   - Adds the request builders for Google Login and Join. 
   - Converts `marketingOptIn` to String before adding them into the request parameter dictionary

#### How to Test

- Confirm unit tests pass
- See client PR for testing that authentication flow still works

### CODE REVIEW

**Comment by nicolelehrer** [comment]
@rpassis I had success passing a bool instead of String - I'll confirm with the API team when I'm back and if either work, I'll remove converter 👍

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">SwiftLint not installed, download from https://github.com/realm/SwiftLint</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 118 tests, with 0 failures (0 unexpected) in 1.502 (1.787) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoNetworking/1421)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Sources/Shared/AuthenticationController.swift [hunk 1]
```diff
@@ -291,17 +292,55 @@ final public class AuthenticationController {
...
     /**
      **(PRIVATE: Vimeo Use Only, will not work for third-party applications)**
-     Join with a facebook token
+     Join with a Facebook token
      
      - parameter facebookToken: token from facebook SDK
      - parameter completion:    handler for authentication success or failure
      */
-    public func facebookJoin(withToken facebookToken: String, marketingOptIn: String, completion: @escaping AuthenticationCompletion) {
+    public func facebookJoin(withToken facebookToken: String, marketingOptIn: Bool, completion: @escaping AuthenticationCompletion) {
```
Comment (rpassis) [inline Sources/Shared/AuthenticationController.swift:302]: can we fix the documentation here to include the `marketingOptIn` parameter?

File: Sources/Shared/Request+Authentication.swift [hunk 2]
```diff
@@ -36,7 +36,8 @@ private let UsernameKey = "username"
 private let PasswordKey = "password"
```
Comment (rpassis) [inline Sources/Shared/Request+Authentication.swift]: What do you think of moving all of these constants to the bottom of this file so they are out of the way?

File: Sources/Shared/Request+Authentication.swift [hunk 3]
```diff
@@ -130,8 +134,8 @@ extension Request where ModelType: VIMAccount {
      
      - returns: a new `Request`
      */
-    static func joinRequest(withName name: String, email: String, password: String, marketingOptIn: String, scopes: [Scope]) -> Request {
```
Comment (rpassis) [inline Sources/Shared/Request+Authentication.swift:133]: can we update documentation with the `marketingOptIn` parameter

File: Sources/Shared/Request+Authentication.swift [hunk 4]
```diff
@@ -164,14 +168,49 @@ extension Request where ModelType: VIMAccount {
...
+        ]
+        
+        return Request(method: .POST, path: AuthenticationPathGoogleToken, parameters: parameters)
+    }
+    
+    /// Constructs a `Request` for joining with Google. For internal use only.
+    ///
+    /// - Parameters:
+    ///   - googleToken: `idToken` returned by the GoogleSignIn SDK
+    ///   - marketingOptIn: bool indicateing whether a user has opted-in to receive marketing material
```
Comment (rpassis) [inline Sources/Shared/Request+Authentication.swift]: typo on indicating here

**Review by rpassis — CHANGES_REQUESTED** [review]
Hey only some minor comments that I am releasing while I test the changes on a device.

File: Sources/Shared/Request+Authentication.swift (hunk 2 above)
Comment (nicolelehrer) [reply Sources/Shared/Request+Authentication.swift]: I'm down - I moved only the private constants to the bottom and left the public typealias on top if that sounds good

**Review by rpassis — APPROVED** [review]
LGTM 🚀

**Pull request merged**